controller/lib/hal/uart_dma.h

```diff
@@ -83,7 +59,8 @@ class UART_DMA {
   // setup. Returns true if no reception is in progress and new reception
   // was setup.
 
-  bool startRX(const char *buf, uint32_t length, uint32_t timeout);
+  [[nodiscard]] bool startRX(const uint8_t *buf, uint32_t length,
```

Comment:
The comment seems wrong - it is reception of **at most** length chars, right?

Reply:
Actually it is **exactly** length chars  :)

Reply:
Please edit the comment then.

I'm still not completely following how this works though. So, the line is idle, and then:
* If it stays idle for timeout, then it's a timeout error (and we call rx->OnRxError - please mention that in the comment)
* If it becomes active, then we start receiving data
* If we receive exactly length bytes, then it's good (and then what happens? We call rxl->OnRxComplete, right? Please say so in the comment)

But this leaves a few questions open:
* After the line is not idle, how long do we wait for receiving length bytes? Is it also "timeout", and if so, does it count from the moment you called StartRx, or from the moment the line stopped being idle?
* What happens if we receive less than length bytes within that timeframe?
* What happens if we receive more than length bytes within that timeframe? If we just receive the first length - what happens to the rest: do they count towards the next rx, or are they simply discarded?

Reply:
The comment is updated, sorry, didn't mention that. 

All those events are independent. We order the DMA controller to receive x number of bytes and it will stay receiving those bytes, when DMA transfers the ordered number of bytes it will call OnRxComplete. Character match event is not related to DMA transfer, whenever UART sees the character it will issue the OnCharMatch, DMA transfer is not affected in any way. Similarly with errors - UART will issue an error interrupt calling OnRxError whenever there is an error on the RX line (framing, overflow, timeout) DMA transfer will hang there and wait for bytes, and UART will also keep providing either bytes or error interrupts if those occur (except for timeout - that is reset after a stop bit, i.e. after each successfully received char).

To answer your questions:
 * If line stays idle for timeout after last stop bit received, `OnRxError(TIMEOUT)` is called, that's it, DMA is still waiting for ordered number of bytes to come, driver does not change the reception configuration. This does not mean that transfer is cancelled.
 * Last stop bit is the start of timeout window. I.e. timeout counter starts after last received character. I think I understood what bugs you - timeout is not for the full length ordered for reception, but for the rx line. I.e. if line stays idle for timeout, interrupt is issued. Timeout error is not bound to DMA in any way, it's there to inform the caller that the line is idle. 

It's not clear how useful having the timeout interrupt really is, because we can timestamp frames in Comms decoded by FrameDecoder and notice that we didn't have a frame for too long. I suggest we land the FrameDecoder and if we see that timeout is not used in a reasonable way, I will remove it from UART_DMA.